resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocatorBlacklistTracker.scala

```diff
@@ -47,16 +48,29 @@ import org.apache.spark.util.{Clock, SystemClock}
 private[spark] class YarnAllocatorBlacklistTracker(
     sparkConf: SparkConf,
     amClient: AMRMClient[ContainerRequest],
+    yarnClient: YarnClient,
     failureTracker: FailureTracker)
   extends Logging {
 
+  private var allNodes = Map.empty[String, NodeState]
```

Comment:
This is still not needed as `refreshYarnNodes()` is used only in one place. 
Although it would be nice to have a separate member val for the config read `sparkConf.get(YARN_EXCLUDE_NODES).toSet`. We could call it like `configuredExcludeNodes`.

Regarding `refreshYarnNodes()`my suggestion is get rid of it completely and within `excludeNodes` you could do sth like this (assuming you introduced `configuredExcludeNodes`):

```scala
  val allCurrentlyAvailableNodes =
    yarnClient.getNodeReports().asScala.map(node => (node.getNodeId.getHost, node.getNodeState)).toMap
  val (existentNodes, nonexistentNodes) =
      configuredExcludeNodes.partition(node => allCurrentlyAvailableNodes.contains(node))
  ...
```

Reply:
+1 to dropping `refreshYarnNodes()` -- maybe this is paranoid, but it seems like it would add a lot of RPC volume to the active RM if each Spark app were calling it each time it needed to allocate